### ISSUE TITLE
[notice] CDK CLI Authentication Issues

### ISSUE THREAD

**Issue by rix0rrr** [aws/aws-cdk#1656]
The CDK CLI has some limitations compared to the AWS CLI. If you are having authentication issues, they might be covered by one of the following issues. Please use this thread to discuss further.

## CDK CLI will not read your region from your [default] profile

If your `~/.aws/config` is set up like this:

```
[default]
region = us-east-1

[profile MyProfile]
role_arn = arn:aws:iam::123456789012:role/myprofile
source_profile = SomeOtherProfile
```

Even though your profile `MyProfile` does not contain a `region`, the AWS CLI will read the region from the `[default]` section. The AWS SDK for JavaScript that the CDK is built on does not do this, so the CDK does not support this. Make sure every profile section contains the `region`.

## Cannot have a profile named "default" in the config file

The following will not work:

```
[default]
...

[profile default]
...
```

The AWS CLI seems to accept this, but the AWS SDK for JavaScript will fail to load the configuration file properly.

## How to properly select AWS CLI profiles for your stacks/accounts
_See issue #3961_

**Comment by vschumaker** [comment]
I ran in to the case where a `[profile default]` of the following format

```
[default]
output = text
region = us-east-1

[profile default]
role_arn = arn:aws:iam::123456789012:role/myrole
source_profile = default

[profile dev]
source_profile = default
role_arn = arn:aws:iam::987654321012:role/myrole
region = us-east-1
```

caused the following error, even though my `AWS_PROFILE` environment variable was set to `dev`

```
cdk deploy
Need to perform AWS calls for account unknown-account, but no credentials found. Tried: default credentials.
```

Removing the `[profile default]` section allowed me to successfully `cdk deploy` my stack.

**Comment by moofish32** [comment]
Answering my own question for the next person.

My issue was that at some point `~/.aws/config` added a similar file to the example above. I deleted the `~/.aws/config` because I no longer use it and set the `AWS_DEFAULT_REGION=xxxx` env var in my shell. After that everything started working. @rix0rrr actually found this. So if you have `~/.aws/credentials` and `~/.aws/config` they both can't have a `default` profile section.

**Comment by 0xdevalias** [comment]
Another edge case, in case anyone else runs into it. Using [`aws-vault`](https://github.com/99designs/aws-vault)

```
⇒  aws-vault exec myprofile -- cdk diff
Need to perform AWS calls for account unknown-account, but no credentials found. Tried: default credentials.
```

Looking at `~/.aws/config` I had an empty profile: `[profile myprofile]`

By changing it to include a region, the issue was resolved and `cdk diff` worked as expected:

```
[profile myprofile]
region=ap-southeast-2
```

**Comment by tunagami** [comment]
Just to add on for MFA:

I am having this problem WITHOUT a `[profile default]` entry.  Any idea why I am getting the same error?

```
[default]
region = eu-central-1

[profile int-server]
source_profile = default
role_arn = arn:aws:iam::121212121212:role/deployer
mfa_serial = arn:aws:iam::421212121212:mfa/me@mfa.eu

[profile prod-server]
source_profile = default
role_arn = arn:aws:iam::321212121212:role/deployer
mfa_serial = arn:aws:iam::421212121212:mfa/me@mfa.eu
```

error:

```
Setting "aws:cdk:toolkit:default-region" context to eu-central-1
Resolving default credentials
Unable to determine the default AWS account (did you configure "aws configure"?): { AccessDenied: Access denied
```

**Comment by LechuckThePirate** [comment]
Hello... I'm having another authentication issue... 

We use a identity account and a governor account role with MFA enabled... my config file looks like this:

[profile default]
region=eu-west-1
role_arn=arn:aws:iam::xxxxxxxxxxxxx:role/xxxxxxxx
mfa_serial=arn:aws:iam::xxxxxxxxxxxx:mfa/xxxxxxxxx@xxxxxx.com
source_profile=default

and I have a [default] credentials with my client and secret keys.

I'm running CDK from a AWS Workspace (I don't know if this can be the problem). I get an error saying that I didn't do aws configure ... I did, and I'm providing --profile default to be sure.

Any fix/workaround?

If I try to deploy the hello-cdk app I get this (with --verbose)

```
PS > cdk deploy --profile default -v
CDK toolkit version: 0.27.0 (build 7638a58)
Command line arguments: { _: [ 'deploy' ],
  trace: false,
  strict: false,
  'ignore-errors': false,
  ignoreErrors: false,
  json: false,
  j: false,
  verbose: true,
  v: true,
  ec2creds: undefined,
  i: undefined,
  'version-reporting': undefined,
  versionReporting: undefined,
  'path-metadata': true,
  pathMetadata: true,
  'asset-metadata': true,
  assetMetadata: true,
  ci: false,
  version: false,
  help: false,
  h: false,
  exclusively: false,
  e: false,
  profile: 'default',
  'role-arn': undefined,
  r: undefined,
  roleArn: undefined,
  'build-exclude': [],
  E: [],
  buildExclude: [],
  '$0': '..\\..\\..\\AppData\\Roaming\\npm\\node_modules\\aws-cdk\\bin\\cdk',
  app: undefined,
  context: undefined,
  plugin: undefined,
  rename: undefined,
  proxy: undefined,
  'toolkit-stack-name': undefined,
  'require-approval': undefined,
  STACKS: [] }
Determining whether we're on an EC2 instance.
cdk.json: {
  "app": "node bin/testapp.js"
}
Setting "aws:cdk:toolkit:default-region" context to undefined
Resolving default credentials
Looks like EC2 instance.
Looking up default account ID from STS
Unable to determine the default AWS account (did you configure "aws configure"?): { InvalidClientTokenId: The security token included in the request is invalid.
    at Request.extractError (D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\node_modules\aws-sdk\lib\protocol\query.js:50:29)
    at Request.callListeners (D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\node_modules\aws-sdk\lib\sequential_executor.js:106:20)
    at Request.emit (D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\node_modules\aws-sdk\lib\sequential_executor.js:78:10)
    at Request.emit (D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\node_modules\aws-sdk\lib\request.js:683:14)
    at Request.transition (D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\node_modules\aws-sdk\lib\request.js:22:10)
    at AcceptorStateMachine.runTo (D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\node_modules\aws-sdk\lib\state_machine.js:14:12)
    at D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\node_modules\aws-sdk\lib\state_machine.js:26:10
    at Request.<anonymous> (D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\node_modules\aws-sdk\lib\request.js:38:9)
    at Request.<anonymous> (D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\node_modules\aws-sdk\lib\request.js:685:12)
    at Request.callListeners (D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\node_modules\aws-sdk\lib\sequential_executor.js:116:18)
  message: 'The security token included in the request is invalid.',
  code: 'InvalidClientTokenId',
  time: 2019-04-01T10:23:27.706Z,
  requestId: '30626f19-5468-11e9-b302-4524a75e3687',
  statusCode: 403,
  retryable: false,
  retryDelay: 16.893920433979147 }
Setting "aws:cdk:toolkit:default-account" context to undefined
context: { 'aws:cdk:toolkit:default-region': undefined,
  'aws:cdk:toolkit:default-account': undefined,
  'aws:cdk:enable-path-metadata': true,
  'aws:cdk:enable-asset-metadata': true }
outdir: D:\xxxxxxxxxxxxxx\AppData\Local\Temp\cdkYL2v9X
outfile: D:\xxxxxxxxxxxxxx\AppData\Local\Temp\cdkYL2v9X\cdk.out
{ version: '0.19.0',
  artifacts: 
   { TestappStack: 
      { type: 'aws:cloudformation:stack',
        environment: 'aws://unknown-account/unknown-region',
        properties: [Object],
        metadata: [Object] } },
  runtime: 
   { libraries: 
      { '@aws-cdk/cdk': '0.27.0',
        '@aws-cdk/cx-api': '0.27.0',
        '@aws-cdk/aws-s3': '0.27.0',
        '@aws-cdk/aws-events': '0.27.0',
        '@aws-cdk/aws-iam': '0.27.0',
        '@aws-cdk/region-info': '0.27.0',
        '@aws-cdk/aws-kms': '0.27.0',
        '@aws-cdk/aws-s3-notifications': '0.27.0',
        '@aws-cdk/aws-codepipeline-api': '0.27.0',
        'jsii-runtime': 'node.js/v8.12.0' } },
  stacks: 
   [ { name: 'TestappStack',
       environment: [Object],
       template: [Object],
       metadata: [Object] } ] }
Removing outdir D:\xxxxxxxxxxxxxx\AppData\Local\Temp\cdkYL2v9X
The stack TestappStack already includes a CDKMetadata resource
Stack name not specified, so defaulting to all available stacks: TestappStack
Reading existing template for stack TestappStack.
Need to perform AWS calls for account unknown-account, but no credentials found. Tried: default credentials.
Error: Need to perform AWS calls for account unknown-account, but no credentials found. Tried: default credentials.
    at CredentialsCache.getCredentials (D:\xxxxxxxxxxxxxx\AppData\Roaming\npm\node_modules\aws-cdk\lib\api\util\sdk.ts:209:11)
    at <anonymous>

```

**Comment by Sparkboxx** [comment]
**CDK not using the `--profile` setting**

Is it expected behaviour that when a `profile` is configured in `cdk.json` that  `cdk` uses your `[default]` credential profile if you don't give it an explicit `--profile` ? Or should it be safe to assume that if the profile is set in the `cdk.json` that when running a `cdk` command it uses the configured profile?

**The story**
I forgot to include the `--profile myProfile` a bunch of times while creating my app and ended up creating infrastructure my `[default]` profile account. This of course is user error, but I think an error other people might make too.

In order to avoid the problem I now setup an empty `[default]` profile, so that if I forget to include `--profile myProfile` I get an error. (Which is probably a good thing for me to do anyway working with multiple profiles)

CDK is configured with:

```
{
  "app": "node bin/app.js",
  "profile": "myProfile"
}
```

**Comment by kadishmal** [comment]
I'm trying to deploy a CDK app from within a Lambda function. Now it complains with:

    Need to perform AWS calls for account unknown-account, but no credentials found. Tried: default credentials.

Passing `--role-arn` doesn't seem to work. Why is CDK bound to `~/.aws/credentials` and `~/.aws/config`?

Lambda function itself already has a role that has all the permissions to deploy a CDK app. Is this even possible? The final resort is to duplicate CDK's `cdk.js` and remove the credential checker part which feels so unnecessary.

**Comment by mrgrain** [comment]
I have a CDK app which is deployed from a docker container inside an EC2 instance.

The EC2 instance has a role in the target account which does nothing but allow me to assume an other role.

I have a CDK plugin which handles assuming the role.

When I try to deploy via CDK, the app is loading the plugin, but is not requesting credentials from it. Instead CDK resolves and looks up "default credentials" for the account. Since the role is for the account, CDK moves on and subsequently fails as the instance role has no permissions.

Is there a way to tell CDK to not use this account, or better to just use my plugin? 

`AWS_EC2_METADATA_DISABLED` won't work as the plugin is relying on metadata.

**Comment by kadishmal** [comment]
@mrgrain I went on with a different approach I described in this comment https://github.com/awslabs/aws-cdk/issues/2637#issuecomment-496291068. It might give you some ideas.

**Comment by mrgrain** [comment]
@kadishmal I found a solution for my issue. Since the problem came down to AWS CDK trying to use ec2 meta data credentials by default, we can disable that behaviour by setting `--ec2creds false` on the cli.

I guess what would be nice if that option is explicitly called out in the log. Something like
```
Looks like an EC2 instance. Trying ec2 meta data credentials (disable this behaviour with `--ec2creds false`).
```

**Comment by zzenonn** [comment]
I have two weird issues:

My config file is setup like this:

```
[default]
output = json
region = ap-southeast-1
role_arn = arn:aws:iam::123456789012:role/AdminRole
source_profile = default

[profile company]
region=ap-southeast-1
output=json
cli_follow_urlparam = false

[profile educate]
region=us-east-1
output=json

```

When I run `cdk deploy`, 

My first issue is I get this error which uses a different region from my default (which is very strange). 

My second issue is that the CDK also does not assume the role specified in the configuration.

```
User: arn:aws:iam::123456789012:user/zenon is not authorized to perform: cloudformation:GetTemplate on resource: arn:aws:cloudformation:us-east-2:123456789012:stack/hello-cdk-1/*
```

**Comment by mbonig** [comment]
Is there any verified way of leverage just the AWS_SECRET_ACCESS_KEY and AWS_ACCESS_KEY_ID (but no AWS_SESSION_TOKEN) environment variables?

**Comment by skinny85** [comment]
> Is there any verified way of leverage just the AWS_SECRET_ACCESS_KEY and AWS_ACCESS_KEY_ID (but no AWS_SESSION_TOKEN) environment variables?

I believe if you're not using temporary credentials like assumeRole, but something like an IAM User, you don't need to set AWS_SESSION_TOKEN.

**Comment by kevinslin** [comment]
for mfa, I use [aws-mfa]( https://github.com/broamski/aws-mfa) to make things a bit more bearable. it writes short term credentials into your .credentials file and works with cdk!

**Comment by mbonig** [comment]
> > Is there any verified way of leverage just the AWS_SECRET_ACCESS_KEY and AWS_ACCESS_KEY_ID (but no AWS_SESSION_TOKEN) environment variables?
> 
> I believe if you're not using temporary credentials like assumeRole, but something like an IAM User, you don't need to set AWS_SESSION_TOKEN.

So, just to round out the issue... I have been trying to run `cdk synth` inside a Jenkins pipeline that was given temp credentials from the AWS Credentials Jenkins plugin. However, CDK couldn't find the env variable creds properly. So, I just write them to a file first and then point AWS_SHARED_CREDENTIALS_FILE at it and now things are working fine.

**Comment by jonnyyu** [comment]
1. using aws-cdk 1.3.0 on a EC2 AMI Linux 2 instance with role attached.
2. run 'aws configure' with secret/access=None, region=us-west-2
3. run 'cdk deploy myservice' failed with 
```
Need to perform AWS calls for account XXXXXXXXXXX, but no credentials found. 
```
Detail logs:
```
Tried: default credentials.
Error: Need to perform AWS calls for account XXXXXXXXXXX, but no credentials found. Tried: default credentials.
    at CredentialsCache.getCredentials (/home/ec2-user/.nvm/versions/node/v12.7.0/lib/node_modules/aws-cdk/lib/api/util/sdk.ts:261:11)
    at processTicksAndRejections (internal/process/task_queues.js:85:5)
    at CredentialsCache.get (/home/ec2-user/.nvm/versions/node/v12.7.0/lib/node_modules/aws-cdk/lib/api/util/sdk.ts:223:25)
    at SDK.cloudFormation (/home/ec2-user/.nvm/versions/node/v12.7.0/lib/node_modules/aws-cdk/lib/api/util/sdk.ts:117:20)
    at CloudFormationDeploymentTarget.readCurrentTemplate (/home/ec2-user/.nvm/versions/node/v12.7.0/lib/node_modules/aws-cdk/lib/api/deployment-target.ts:52:17)
    at CdkToolkit.deploy (/home/ec2-user/.nvm/versions/node/v12.7.0/lib/node_modules/aws-cdk/lib/cdk-toolkit.ts:94:33)
    at main (/home/ec2-user/.nvm/versions/node/v12.7.0/lib/node_modules/aws-cdk/bin/cdk.ts:192:16)
    at initCommandLine (/home/ec2-user/.nvm/versions/node/v12.7.0/lib/node_modules/aws-cdk/bin/cdk.ts:150:9)
```
A workaround is to `touch ~/.aws/credentials`, then it can correctly retrieve credentails from EC2 metadata.
But AWS CLI doesn't require `~/.aws/credentials` exist

**Comment by SteveHoggNZ** [comment]
@kadishmal I got CDK running in a Lambda. Setting the a HOME environment variable to /tmp for the Lambda allows create the credentials files and use the Lambda's role.

**Comment by kadishmal** [comment]
@SteveHoggNZ just like that? Set the HOME env to /tmp and CDK will stop using the credentials? Can you paste some sample code?

**Comment by whilp** [comment]
> A workaround is to touch ~/.aws/credentials, then it can correctly retrieve credentials from EC2 metadata.

This workaround may feel slightly less painful if you use the `AWS_SHARED_CREDENTIALS_FILE` environment variable. 

```bash
export AWS_SHARED_CREDENTIALS_FILE=/tmp/aws-cdk-1656-workaround
touch $AWS_SHARED_CREDENTIALS_FILE
```

**Comment by helzich** [comment]
Adding to the findings of @jonnyyu, AWS CDK does not only expect `~/.aws/credentials`  to exist, it also requires the profile information to reside in `~/.aws/config` even if the environment variable `AWS_CONFIG_FILE` points elsewhere. The AWS CLI respects the environment variable, but AWS CDK does not (tested with 1.5.0 (build c020efa) in a docker container running "Amazon Linux 2").

**Comment by konstantinj** [comment]
Hi, I'm wondering how to have stacks of different accounts/regions in one app and synth them at the same time.
Similar to #1035 and #290

I'm specifying accountId and region with `cdk.StackProps.env` but (of course) cdk doesn't have access to all accounts specified:

`Need to perform AWS calls for account XXXXXXXXXXXX, but no credentials found. Tried: default credentials.`

How is this supposed to work? @eladb @rix0rrr 

For using aws cli I have my default profile that has access to our main account with access/secret keys and all other accounts have then cross-account roles.

**Comment by konstantinj** [comment]
looks like #3961 will solve this issue

**Comment by NGL321** [comment]
_Appending other pinned auth issue here:
#3961_

**Comment by hupe1980** [comment]
We've solved our authentication issues with a simple plug-in that adds support for mfa and multi profiles/accounts: https://github.com/hupe1980/cdk-multi-profile-plugin

**Comment by toleabivol** [comment]
I had this issue that `--profile` paramater had no effect (always getting the default) and it was because of AWS system variables being set. Removed `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` system vars and it worked.

**Comment by holmesjr** [comment]
CDK doesn't support CLI v2 SSO.

**Comment by mitchlloyd** [comment]
The CDK credentials provider plugin continues to be the best option for my team because, as @hupe1980 showed, it is very flexible.

The two issues with the current plugin architecture I run into are:
1. The env vars and default profile always take precedence over the plugin. People sometimes forget or don't realize that they have env vars or a default profile and have to troubleshoot. I wish I could fully delegate credentials work to the plugin.
2. There is no good way to set the role to use for sts in the plugin. We assume a conventional role name (`Admin`), but sometimes teams use different role names and need to provide additional info in an environment variable. It would be nice if the stack `env` had an optional `role` parameter to pass to the plugin, but maybe there's a better place for this.

In practice these 2 issues aren't too bad if CDK is the primary way developers interact with AWS. However the myriad of other AWS credentials processes derail things.

**Comment by dcoty** [comment]
Regarding MFA support, There are a couple of possible workarounds, either can be scripted to make life easier:

- Follow this: https://aws.amazon.com/premiumsupport/knowledge-center/authenticate-mfa-cli/ and manually update your credentials file, then use CDK. This was suggested by Amazon Support.

- Use python aws-mfa library( i made some instructions at https://github.com/aws/aws-cdk/issues/1248#issuecomment-571355271) This basically does the same thing as above but 
 takes less steps. 

In the end, these are just workarounds, and it would be nice to get this into the product.

**Comment by lrhazi** [comment]
FYI, I was getting this error:
```
Need to perform AWS calls for account 9999-9999-9999, but no credentials found. Tried: default credentials.
```
and it turns out the dashes in the account number were the problem!

**Comment by skorfmann** [comment]
> @kadishmal I found a solution for my issue. Since the problem came down to AWS CDK trying to use ec2 meta data credentials by default, we can disable that behaviour by setting `--ec2creds false` on the cli.
> 
> I guess what would be nice if that option is explicitly called out in the log. Something like
> 
> ```
> Looks like an EC2 instance. Trying ec2 meta data credentials (disable this behaviour with `--ec2creds false`).
> ```

This worked for us 💯  in a CI context on EC2 instances. 👍 for making this obvious in the log output, since it's not obvious that the metadata endpoint will have priority over credentials in the env. Comparing this [to the Javascript SDK](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-credentials-node.html):

```
If more than one credential source is available to the SDK, the default precedence of selection is as follows:

- Credentials that are explicitly set through the service-client constructor
- Environment variables
- The shared credentials file
- Credentials loaded from the ECS credentials provider (if applicable)
- Credentials that are obtained by using a credential process specified in the shared AWS config file or the shared credentials file. 
- Credentials loaded from AWS IAM using the credentials provider of the Amazon EC2 instance (if configured in the instance metadata)
```

So, even better than logging and documentation this behaviour as it is right now, it would probably better to align it with the rest of the ecosystem.

**Comment by cn-sws** [comment]
I was getting either of these two errors depending on what I tried in the config/credentials files to get it working:
> Unable to determine AWS region from environment or AWS configuration

> Unable to determine the default AWS account (did you configure "aws configure"?)

In my case, the cause had nothing to do with the profile I was actually trying to use, i.e. --profile sbox-03.  The cause was from a "pseudo" duplicate entry in the config file. I had these two entries:
```
[app]
region = ap-southeast-2
output = json
[profile app]
region = ap-southeast-2
output = json
```
This caused the error below when running this command:
```
cdk deploy -v --profile sbox-03
```
Note the "-v".  Without the -v you just get a one line response:
> Unable to resolve AWS account to use. It must be either configured when you define your CDK or through the environment

the error:
```
Unable to determine the default AWS account (did you configure "aws configure"?): TypeError: Cannot redefine property: app
    at Function.defineProperty (<anonymous>)
    at D:\Data\CIT\sw projects\_app-support\node-v10.12.0-win-x64\node_modules\aws-cdk\node_modules\aws-sdk\lib\shared-ini\ini-loader.js:11:14
    at Array.forEach (<anonymous>)
    at IniLoader.parseFile (D:\Data\CIT\sw projects\_app-support\node-v10.12.0-win-x64\node_modules\aws-cdk\node_modules\aws-sdk\lib\shared-ini\ini-loader.js:8:26)
    at IniLoader.loadFrom (D:\Data\CIT\sw projects\_app-support\node-v10.12.0-win-x64\node_modules\aws-cdk\node_modules\aws-sdk\lib\shared-ini\ini-loader.js:56:30)
    at Object.getProfilesFromSharedConfig (D:\Data\CIT\sw projects\_app-support\node-v10.12.0-win-x64\node_modules\aws-cdk\node_modules\aws-sdk\lib\util.js:960:42)
    at SharedIniFileCredentials.load (D:\Data\CIT\sw projects\_app-support\node-v10.12.0-win-x64\node_modules\aws-cdk\node_modules\aws-sdk\lib\credentials\shared_ini_file_credentials.js:96:31)
    at SharedIniFileCredentials.coalesceRefresh (D:\Data\CIT\sw projects\_app-support\node-v10.12.0-win-x64\node_modules\aws-cdk\node_modules\aws-sdk\lib\credentials.js:205:12)
    at SharedIniFileCredentials.refresh (D:\Data\CIT\sw projects\_app-support\node-v10.12.0-win-x64\node_modules\aws-cdk\node_modules\aws-sdk\lib\credentials\shared_ini_file_credentials.js:168:10)
    at SharedIniFileCredentials.get (D:\Data\CIT\sw projects\_app-support\node-v10.12.0-win-x64\node_modules\aws-cdk\node_modules\aws-sdk\lib\credentials.js:122:12)`
```
This code in 'ini-loader.js' generates the error:
```
Object.defineProperty(tmpContent, profileName, {
```
```
function parseFile(filename, isConfig) {
    var content = AWS.util.ini.parse(AWS.util.readFileSync(filename));
    var tmpContent = {};
    Object.keys(content).forEach(function(profileName) {
      var profileContent = content[profileName];
      profileName = isConfig ? profileName.replace(/^profile\s/, '') : profileName;
      Object.defineProperty(tmpContent, profileName, {
        value: profileContent,
        enumerable: true
      });
    });
    return tmpContent;
}
```
Notice, it is trying to redefine my "app" profile:
> Cannot redefine property: app

Notice above I said "pseudo" duplicate.  If you have an actual duplicate, i.e. [app] and [app] for profile names then there is no problem.  It is just when you have [app] and [profile app].

**Comment by thekevinbrown** [comment]
I'm struggling with this because the behaviour I want is as follows:

- In CI, use environment variables
- In dev, use [.env](https://www.npmjs.com/package/dotenv)
- Always pull credentials from these sources regardless of what's in the developer's `~/.aws` directory

Why? Because we work on many clients who have their own AWS accounts, and I want to be sure that we never deploy one client's setup to the wrong AWS account, especially in the case where there's a developer who has a `[default]` profile set to a different client but is doing something on this project for a bit.

How can I ensure this happens? I feel like the easiest API here would be to be able to pass in the Access Key and Secret Key directly to the tool and have that always override whatever's in the various files and env vars.

**Comment by john-tipper** [comment]
@SteveHoggNZ how did you package up the CDK into the lambda please?  Did you install as a layer and if so, how did you create this layer?

**Comment by 0xdevalias** [comment]
@thekevinbrown I use https://github.com/99designs/aws-vault to store my credentials, and give each profile within it a meaningful name (eg. client name). Then in my package.json scripts I prefix all of the cdk/etc commands with aws-vault syntax for the relevant client. That way I never have to think about which client, or the differences, I just need to run the same script in each project (and have an appropriate profile for that client configured on my computer, if not, it will fail rather than deploy to the wrong account)

**Comment by SteveHoggNZ** [comment]
> @SteveHoggNZ how did you package up the CDK into the lambda please? Did you install as a layer and if so, how did you create this layer?

@john-tipper we did package CDK in a Lambda. We use Python, so we also had to bundle in NodeJS in the Lambda. Here is a simplified version of what we did:

# Dockerfile
```
FROM lambci/lambda:build-python3.7
 
ARG CDK_VERSION
 
# CDK requires NodeJS
RUN yum -y update && yum install -y gcc-c++ make && \
        curl -sL https://rpm.nodesource.com/setup_12.x | bash - && \
        yum install -y nodejs
 
RUN npm i -g aws-cdk@$CDK_VERSION
 
RUN mkdir -p /opt/bin /opt/lib && \
        cp -pR /usr/lib/node_modules /opt/lib && \
        cp -p /usr/bin/node /opt/bin && \
        strip /opt/bin/node && \
        ln -s /opt/lib/node_modules/npm/bin/npm-cli.js /opt/bin/npm && \
        sed -i 's|/usr/bin/env node|/opt/bin/node|' /opt/lib/node_modules/aws-cdk/bin/cdk
 
# CDK requires the AWS CLI
RUN pip install awscli --target /opt/python/lib/python3.7/site-packages
```

# Docker Build
```
CDK_VERSION=v1.25.0
docker build layers/cdk --tag cdk-layer-builder:$CDK_VERSION --build-arg CDK_VERSION=$CDK_VERSION
```

# Layer Build
```
mkdir -p layers/cdk/versions
docker run \
        -v ${PWD}/layers/cdk/versions:/versions \
        cdk-layer-builder:$CDK_VERSION \
        bash -c "cd /opt && zip -r /versions/cdk-$CDK_VERSION.zip . >/dev/null"
```

That will give you a zip file in layers/cdk/versions you can upload as a layer.

To get the Lambda to run CDK there are other requirements:

- Set the HOME environment variable of the Lambda to /tmp
- When running CDK in the Lambda:
  - append /opt/lib/node_modules/aws-cdk/bin and opt/python/lib/python3.7/site-packages/bin to the PATH environment variable
  - get CDK to output to the tmp folder i.e. `--output /tmp/cdk.out`

That should be it.

My boss has given me the go-ahead to do a blog article on this with more detail.

**Comment by john-tipper** [comment]
Thanks @SteveHoggNZ, particularly for pointing out the gotchas!  I ended up coming up with a similar solution for packaging the CDK as a layer: https://stackoverflow.com/a/60399178/272023

**Comment by FreeWillaert** [comment]
> FYI, I was getting this error:
> 
> ```
> Need to perform AWS calls for account 9999-9999-9999, but no credentials found. Tried: default credentials.
> ```
> 
> and it turns out the dashes in the account number were the problem!

Same here. Having a wrong/missing/bad value in the env.account property of StackProps causes this same error message. Very confusing.

**Comment by misk0** [comment]
I'm having similar issues and after reading whole thread I couldn't do an initial deploy. My setup:

> aws --version
> File association not found for extension .py
> aws-cli/1.16.310 Python/3.8.1 Windows/10 botocore/1.13.46

> cdk --version
> 1.27.0 (build a98c0b3)

I created simple Java class following the documentation and I'm getting 

> cdk deploy Stack-east-1
> Need to perform AWS calls for account testuser, but no credentials found. Tried: default credentials.
> 

 
If I try to use "aws s3 ls" if works without any problem. 

This is verbose output:

> >cdk deploy Stack-east-1 -v
> CDK toolkit version: 1.27.0 (build a98c0b3)
> Command line arguments: {
>   _: [ 'deploy' ],
>   v: true,
>   verbose: true,
>   'ignore-errors': false,
>   ignoreErrors: false,
>   json: false,
>   j: false,
>   ec2creds: undefined,
>   i: undefined,
>   'version-reporting': undefined,
>   versionReporting: undefined,
>   'path-metadata': true,
>   pathMetadata: true,
>   'asset-metadata': true,
>   assetMetadata: true,
>   'role-arn': undefined,
>   r: undefined,
>   roleArn: undefined,
>   staging: true,
>   'no-color': false,
>   noColor: false,
>   fail: false,
>   'build-exclude': [],
>   E: [],
>   buildExclude: [],
>   ci: false,
>   execute: true,
>   force: false,
>   f: false,
>   '$0': 'C:\\Users\\misko\\AppData\\Roaming\\npm\\node_modules\\aws-cdk\\bin\\cdk',
>   STACKS: [ 'Stack-east-1' ],
>   stacks: [ 'Stack-east-1' ]
> }
> Determining whether we're on an EC2 instance.
> cdk.json: {
>   "app": "mvn -e -q exec:java"
> }
> cdk.context.json: {
>   "@aws-cdk/core:enableStackNameDuplicates": "true",
>   "aws-cdk:enableDiffNoFail": "true"
> }
> merged settings: {
>   versionReporting: true,
>   pathMetadata: true,
>   output: 'cdk.out',
>   app: 'mvn -e -q exec:java',
>   context: {},
>   tags: [],
>   assetMetadata: true,
>   toolkitBucket: {},
>   staging: true
> }
> Setting "CDK_DEFAULT_REGION" environment variable to us-east-1
> Resolving default credentials
> Does not look like EC2 instance.
> Retrieved account ID 201547469031 from disk cache
> Setting "CDK_DEFAULT_ACCOUNT" environment variable to 201547469031
> context: {
>   '@aws-cdk/core:enableStackNameDuplicates': 'true',
>   'aws-cdk:enableDiffNoFail': 'true',
>   'aws:cdk:enable-path-metadata': true,
>   'aws:cdk:enable-asset-metadata': true
> }
> outdir: cdk.out
> env: {
>   CDK_DEFAULT_REGION: 'us-east-1',
>   CDK_DEFAULT_ACCOUNT: '201547469031',
>   CDK_CONTEXT_JSON: '{"@aws-cdk/core:enableStackNameDuplicates":"true","aws-cdk:enableDiffNoFail":"true","aws:cdk:enable-path-metadata":true,"aws:cdk:enable-asset-metadata":true}',
>   CDK_OUTDIR: 'cdk.out',
>   CDK_CLI_ASM_VERSION: '1.21.0',
>   CDK_CLI_VERSION: '1.27.0'
> }
> Reading existing template for stack Stack-east-1.
> Need to perform AWS calls for account testuser, but no credentials found. Tried: default credentials.
> Error: Need to perform AWS calls for account testuser, but no credentials found. Tried: default credentials.
>     at CredentialsCache.getCredentials (C:\Users\misko\AppData\Roaming\npm\node_modules\aws-cdk\lib\api\util\sdk.ts:284:11)
>     at processTicksAndRejections (internal/process/task_queues.js:97:5)
>     at CredentialsCache.get (C:\Users\misko\AppData\Roaming\npm\node_modules\aws-cdk\lib\api\util\sdk.ts:246:25)
>     at SDK.service (C:\Users\misko\AppData\Roaming\npm\node_modules\aws-cdk\lib\api\util\sdk.ts:154:20)
>     at CloudFormationDeploymentTarget.readCurrentTemplate (C:\Users\misko\AppData\Roaming\npm\node_modules\aws-cdk\lib\api\deployment-target.ts:57:17)
>     at CdkToolkit.deploy (C:\Users\misko\AppData\Roaming\npm\node_modules\aws-cdk\lib\cdk-toolkit.ts:114:33)
>     at main (C:\Users\misko\AppData\Roaming\npm\node_modules\aws-cdk\bin\cdk.ts:218:16)
>     at initCommandLine (C:\Users\misko\AppData\Roaming\npm\node_modules\aws-cdk\bin\cdk.ts:168:9)
> 

I don't have any environment variables set and only my config file is populated. Where could be the problem? Thanks.

**Comment by jakebanks** [comment]
I was also experiencing this with `aws-vault`. The working combination of calls for me was:

```
const amiId = StringParameter.fromStringParameterAttributes(this, 'amiId', {
      parameterName: '/my-ami-parameter',
    }).stringValue;
```

Then for my ASG:
`machineImage: new ec2.GenericLinuxImage({ [region]: amiId }),` - note the brackets are significant.

**Comment by misk0** [comment]
Thanks @jakebanks this helped. I'm using Java classes and I followed instruction from AWS doc site. So your post gave me an idea that maybe I should not specify credentials but used the default one. 
And bang! it works. That's ok, but Amazon advises to use different profile for different deployment scenario so it should work.
Here are my classes

> package com.myorg;
> 
> import software.amazon.awscdk.core.App;
> import software.amazon.awscdk.core.Stack;
> import software.amazon.awscdk.core.Environment;
> import software.amazon.awscdk.core.StackProps;
> import software.amazon.awscdk.services.s3.Bucket;
> 
> import java.util.Arrays;
> 
> public class SimpleCdkApp {
> 
>     private static App app;
> 
>     // Helper method to declare MyStacks in specific accounts/regions
>     private static SimpleCdkStack makeMyStack(final String name, final String account,
>                                        final String region) {
> 
>         return new SimpleCdkStack(app, name, StackProps.builder()
>                 .env(Environment.builder()
>                         .account(account)
>                         .region(region)
>                         .build())
>                 .build());
>     }
>     public static void main(final String[] args) {
>         app = new App();
> 
>         //makeMyStack("Stack-east-1",   "testuser",   "us-east-1");
> 
>         new SimpleCdkStack(app, "SimpleCdkStack");
> 
>         app.synth();
>     }
> }


and

> 
> package com.myorg;
> 
> import software.amazon.awscdk.core.Construct;
> import software.amazon.awscdk.core.Stack;
> import software.amazon.awscdk.core.StackProps;
> import software.amazon.awscdk.services.s3.Bucket;
> 
> public class SimpleCdkStack extends Stack {
>     public SimpleCdkStack(final Construct scope, final String id) {
> 
>         this(scope, id, null);
>     }
> 
>     public SimpleCdkStack(final Construct scope, final String id, final StackProps props) {
>         super(scope, id, props);
> 
>         // The code that defines your stack goes here
>         Bucket.Builder.create(this, "MyFirstBucket")
>                 .versioned(true).build();
>     }
> }

If I use default stack it works, if I use helper method and specify profile details that it doesn't work. Any ideas? Thanks

**Comment by rcollette** [comment]
The issue of --profile not overriding environment variables in https://github.com/aws/aws-cdk/issues/1656#issuecomment-558218200 by @toleabivol is a real PITA.

The behavior I was seeing was that stacks having a stack as their parent (not nested stacks mind you), would not get the assigned stackName but would appear with the default hashed stack name.

It was only after using the verbose option that I could see that what was actually happening was invalid credentials were being used.   Of course, those credentials don't print out, which is a pain, and it took forever to figure it out by basically removing any other credential on my system except the profile I was using and realizing that it was the environment variable credentials that were being used.

This behavior is unlike the AWS CLI so it's very confusing.

**Comment by evanderkoogh** [comment]
Found another issues. Authentication doesn't work at all if there `~/.aws/credentials` file does not exist, even though you set the environment variables which should take precedence.

`aws-cli` works just fine, the `aws-sdk-js` fails with: `Error: ENOENT: no such file or directory, open '/Users/<username>/.aws/credentials'`, but `cdk` fails with a generic error message.

**Comment by thekevinbrown** [comment]
I don't want to use `aws-vault`, I want to be able to give credentials to the CDK without writing to files or using third party tools.

Does anyone have a solution for this use case?

**Comment by julienlepine** [comment]
Found a possible workaround to get the actual creds from the CLI directly into CDK. (you still need a `.aws/credentials` file, but it can be empty (as referenced by @evanderkoogh). This leverages the profiles, MFA, or SSO from the CLI, and runs CDK with the target environment set.

`aws creds exec cdk deploy`

**Comment by devarpi** [comment]
This is stopping us from cross account deployment. And it has become blocker now.

**Comment by RicoToothless** [comment]
I face another issue. I try cdk `1.32.2` & `1.33.0` version and show error.
```
The security token included in the request is invalid
```

I try those command work fine.
```
aws sts get-caller-identity
aws s3 ls
```

the workaround is use environment
`export AWS_ACCESS_KEY_ID=somekey`
`export AWS_SECRET_ACCESS_KEY=somesecret`

I think is relate `Cloud9` environment. Because I worked fine in local.
But I was work fine on `Cloud9` before. So I am not sure is cdk or `Cloud9` problem.

**Comment by thedancollins** [comment]
> CDK doesn't support CLI v2 SSO.

FWIW I got the AWS CDK to work against an AWS account that was configured to authenticate via SSO.  I am not sure of the exact steps because I have been dancing with this for two days, but what appears to have worked might require that you do not try to use the --profile flag and let the commands run against "default".  Then you need to clear out any Windows Environment Variables, set the .aws/credentials file with the AWS Key ID, Secret Key AND Session Key and then run "aws configure sso" as normal.

**Comment by davidsteed** [comment]
Have been using cdk for some time.  Never had issues before moved to a new machine and literally cannot git cdk working as it says no creddential have been configured when they have . aws command line, amplify etc. everything else works.   This is a bug It needs fixing it is wasting a lot of time for users

**Comment by aisamu** [comment]
> `aws creds exec cdk deploy`

@julienlepine is this available on a future version of the cli?

**Comment by mateja82** [comment]
Not sure if this helps anyone, but in my organization we use AWS SSO, with over 90 AWS Accounts, with MFA, so it was impossible to manage CDK without AWS SSO support. I found a way to solve it, so till its officially relesed, you can use this, it works quite all right: [MatsCloud blog - CDK with AWS SSO multi account multi profile](https://www.matscloud.com/blog/2020/06/25/how-to-use-aws-cdk-with-aws-sso-profiles/)

**Comment by drissamri** [comment]
CDK deploy works for us on GitLab CI until we try to assume roles to deploy to other accounts:
```
Could not load credentials from TemporaryCredentials: Missing credentials in config, if using AWS_CONFIG_FILE, set AWS_SDK_LOAD_CONFIG=1: Could not load credentials from SharedIniFileCredentials: ENOENT: no such file or directory, open '/root/.aws/credentials'
```

**Comment by drissamri** [comment]
After banging my head on this for a few days, finally got it working..

### Usecase
- GitLab CI (but Jenkins, Bitbucket, etc should be the same)
- No aws/credentials or aws/config file
- AWS_SECRET_ACCESS_KEY & AWS_ACCESS_KEY_ID set with a user that can assume roles on other accounts

### Solution
- Assume cross account role and put in variable (KST)
- Create credentials file
- Set AWS_SHARED_CREDENTIALS_FILE to created credentials file
- Unset AWS_SECRET_ACCESS_KEY & AWS_ACCESS_KEY_ID variables which allowed me to do the assuming
```
    - KST=(`aws sts assume-role --role-arn "arn:aws:iam::$AWS_DEPLOY_ACCOUNT:role/$AWS_DEPLOY_ROLE" --role-session-name "$CI_JOB_ID-dev" --query '[Credentials.AccessKeyId,Credentials.SecretAccessKey,Credentials.SessionToken]' --output text`)
    - printf "[default]\naws_access_key_id=${KST[0]}\naws_secret_access_key=${KST[1]}\naws_session_token=${KST[2]}" > credentials
    - export AWS_SHARED_CREDENTIALS_FILE=credentials && unset AWS_ACCESS_KEY_ID && unset AWS_SECRET_ACCESS_KEY
```

Hope this won't be needed in the near future.. If theres an easier way, do let me know here or on Twitter @drissamri88

**Comment by robert-hanuschke** [comment]
that helped a lot, thanks @drissamri !
For those using CodeBuild, be sure to also `export` the KST variable to be usable in the following command.

**Comment by dhensen** [comment]
I'm having the same issue, normally I would not bump this as enough people are troubled by this, but I'm using the **awsprocesscreds-saml** to get access to a corporate account. My config might look like this:

```
[default]
region = eu-west-2

[profile mn-x01-dnp]
region = eu-central-1
credential_process=awsprocesscreds-saml -e https://sts.mn-services.nl/adfs/ls/IdpInitiatedSignOn.aspx?loginToRp=urn:amazon:webservices -u my_email_adress_is_here@companydomain.nl -p adfs -a arn:aws:iam::111111111111:role/mn-developer
```

(I changed my emailaddress and my account id)
aws-cli is working perfectly.

**Comment by dtelaroli** [comment]
Tests with no success (1.70.0):

**_Using --server_**
```
cdk synth -i or 
cdk synth --ec2cred
```

**_Directly_**

```
aws-vault exec my-profile -- cdk synth
```

_**Results**_

```
Cannot retrieve value from context provider ssm since account/region are not specified at the stack level. Either configure "env" with explicit account and region when you define your stack, or use the environment variables "CDK_DEFAULT_ACCOUNT" and "CDK_DEFAULT_REGION" to inherit environment information from the CLI (not recommended for production stacks)
Subprocess exited with error 1
```

Configuring explicitly account and region I received another error:

```
ENAMETOOLONG: name too long, lstat '/Users/dtelaroli/vitrine/cdk/cdk.out/asset.cd1c5b327c1fa83b53cf74ac46abb1f897478a725cba6d1997e654e2fe04b0bf/cdk.out/asset.cd1c5b327c1fa8[...]
```

Downgrading to 1.59:

![image](https://user-images.githubusercontent.com/536500/97191668-de252e80-1785-11eb-8b62-b15fc9ca0001.png)

I'm lost!
Any workaround?

**Comment by hoegertn** [comment]
The screenshot looks like a dependency version mismatch. Remove your lockfile and node_modules folder and install again.

**Comment by dtelaroli** [comment]
You right, but solving it I get again:

```
/cdk.out/asset.901cce368a283e6ec621cec0a2fa44915487e0edac14bb928adb66d5d47f38c2/cdk.out/asset.901cce368a283e6ec621cec0a2fa44915487e0edac14bb928adb66d5d47f38c2/cdk.out/asset.901cce368a283e6ec621cec0a2fa44915487e0edac14bb928adb66d5d47f38c2/cdk.out/asset.901cce368a283e6ec621cec0a2fa44915487e0edac14bb928adb66d5d47f38c2/cdk.out/asset.901cce368a283e6ec621cec0a2fa44915487e0edac14bb928adb66d5d47f38c2/cdk.out/asset.901cce368a283e6ec621cec0a2fa44915487e0edac14bb928adb66d5d47f38c2/cdk.out/asset.901cce368a283e6ec621cec0a2fa44915487e0edac14bb928adb66d5d47f38c2/cdk.out/asset.901cce368a283e6ec621cec0a2fa44915487e0edac14bb928adb66d5d47f38c2/cdk.out/asset.901cce368a283e6ec621cec0a2fa44915487e0edac14bb928adb66d5d47f38c2/cdk.out/asset.3c401e72ea16f4d9e87e74ee10991f1bb706f7689a13c63210555ee59affb287'
Subprocess exited with error 1
```

**Comment by dtelaroli** [comment]
Never mind people, this problem is a bed configuration.
The problem now, it's just with aws-vault permission. It's bed configure account id, because I'm using multiple-accounts.

**Comment by usmanbinnaeem** [comment]
Hi, Hope You all are doing well, need little help, I have created my IAM user and configured every thing fine, but still getting this error, "Unable to resolve AWS account to use. It must be either configured when you define your CDK or through the environment" any body can help to know what I am doing wrong.

here is my .aws/config and .aws/credentials files
.aws/credentials :

[default]
aws_access_key_id=AKIAZGKxxxxxxxxxxx
aws_secret_access_key=aK83fP4v7Lmsly8Yxxxxxxxxxxxxx

.aws/config:

[default]
region=us-east-1

Regards: usman Naeem.

**Comment by rix0rrr** [comment]
This issue was originally intended as an announcement about differences between CDK CLI auth configuration and AWS CLI auth configuration.

It has morphed into a thread where everybody posts an error which they think might have something to do with authentication.

We will move the original content to the developer guide, and close this thread. If you're not sure if the behavior you're seeing is a bug, come and ask on [Slack](https://join.slack.com/t/cdk-dev/shared_invite/zt-mso6p56d-qJp7SOTBvMaQuDrx7R2wHg).

Otherwise, we will start using individual issues to start tracking individual bugs.

**Comment by ebol2000** [comment]
In case anyone is running into this problem while trying to run the script from Octopus Deploy (using this feature: https://octopus.com/docs/deployments/custom-scripts/aws-cli-scripts) please note:

Do not use `--profile`. This seems to get you into this mess. After loosing days monkeying around with alternatives, someone on our team discovered by accident that if you leave off the `--profile` it just worked. After looking at the statement

>
> Precedence of options
> 
> If you specify an option by using one of the environment variables described in this topic, it overrides any value loaded from a profile in the configuration file.
> 
> If you specify an option by using a parameter on the CLI command line, it overrides any value from either the corresponding environment variable or a profile in the configuration file.

from https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html

I realized that Octopus was already setting the environment variables (scoped for session) in the script, and simply by removing the --profile, it worked. Specifically the PS1 script below worked for us 

```
$packageVer = '#{Octopus.Action.Package.PackageVersion}'
Write-Host "packageVer: $packageVer"

Write-Host ""
Set-Location "{insert project path here }"
Write-Host "********************Running Synth********************"
npx cdk synth

Write-Host $"********************Running Deploy********************"
npx cdk deploy --require-approval never
```

**Comment by stepcheunghk** [comment]
After spending a couple hours on this authentication issue, I finally figured this out.

For everyone who has this issue, you may have a try with the following configurations.

What I have done before, it worked perfectly with aws-cli, but **NOT** CDK.

In `~/.aws/credentials`, 
```
[cn]
AWS_ACCESS_KEY_ID=<redacted>
AWS_SECRET_ACCESS_KEY=<redacted>

[global]
AWS_ACCESS_KEY_ID=<redacted>
AWS_SECRET_ACCESS_KEY=<redacted>

[global-dev]
role_arn = arn:aws:iam::<AWS_ACCOUNT_ID>:role/<ASSUME_ROLE_CROSS_ACCOUNT>
source_profile = global
```

CDK returned following error.
`Need to perform AWS calls for account <AWS_ACCOUNT_ID>, but no credentials have been configured`

It seemed CDK cannot read my credentials and profiles.

After tons of attempts, it finally worked with the following configurations.

In `~/.aws/credentials`,

```
[cn]
aws_access_key_id = <redacted>
aws_secret_access_key = <redacted>

[global]
aws_access_key_id = <redacted>
aws_secret_access_key = <redacted>
```

In `~/.aws/config`,
```
[profile global-dev]
role_arn = arn:aws:iam::<AWS_ACCOUNT_ID>:role/<ASSUME_ROLE_CROSS_ACCOUNT>
source_profile = global
```

**Comment by scottwn** [comment]
@rix0rrr the Slack link in this [comment](https://github.com/aws/aws-cdk/issues/1656#issuecomment-800944804) is dead

**Comment by Nintorac** [comment]
*every single bit of software*: works according to aws-cli
aws bit of software: no there are more hoops you must jump in

**Comment by Nintorac** [comment]
Every bit of software: I will login to AWS according to the credentials file because that is the standard
AWS CDK: nnahhh lol too easy..

but seriously, can you just fix this rather than have several issues complaining about it and another issue with the "fix"

**Comment by kizggerg** [comment]
> @kadishmal I found a solution for my issue. Since the problem came down to AWS CDK trying to use ec2 meta data credentials by default, we can disable that behaviour by setting `--ec2creds false` on the cli.
> 
> I guess what would be nice if that option is explicitly called out in the log. Something like
> 
> ```
> Looks like an EC2 instance. Trying ec2 meta data credentials (disable this behaviour with `--ec2creds false`).
> ```

Building on this solution, for anyone who came across this issue when trying to run `aws-vault` in server mode (`aws-vault exec -s <profile>`), passing `--ec2creds true` into your cdk command seems to get CDK to recognize your vault's sts server to obtain its credentials.
